docs/walkthroughs/self-healing.md

```diff
@@ -0,0 +1,223 @@
+
+```bash
+▶ kubectl get envoyconfig kuard
+NAME    NODE ID   ENVOY API   DESIRED VERSION   PUBLISHED VERSION   CACHE STATE
+kuard   kuard     v3          7966f8f4b7        7cbf9788cd          Rollback
+```
+
+We can check that the kuard app is still available at port 8081.
+
+```bash
+▶ curl http://localhost:8081 -v -o /dev/null --silent
+*   Trying ::1:8081...
+* Connected to localhost (::1) port 8081 (#0)
+> GET / HTTP/1.1
+> Host: localhost:8081
+> User-Agent: curl/7.76.1
+> Accept: */*
+>
+* Mark bundle as not supporting multiuse
+< HTTP/1.1 200 OK
+< content-length: 1592
+< content-type: text/html
+< date: Fri, 02 Jul 2021 14:22:36 GMT
+< x-envoy-upstream-service-time: 1
+< server: envoy
+<
+{ [1592 bytes data]
+* Connection #0 to host localhost left intact
+```
+
+Next time a config update is applied and accepted by all proxies, the `Rollback` status will go back to `InSync`.
+
+## **Conclusion**
+
+We have seen how MARIN3R tracks the status of Envoy resource updates and rollbacks the configuration when it sees that a given update is being rejected by all proxies. Its important to note several considerations about the self-healing functionality:
+
+- Failing configuration updates have an impact in Envoy performance because the xDS server will keep trying to send the update to the proxies and the proxies will keep trying to apply the update, failing in a loop. Even though this is mitigated by a backoff retry strategy in the xDS server side, the self-healing functionality avoids this error loop from going on forever. There is one case when this situation is not revertible though, which occurs when there is no previous correct config to rollback to.
+- To mark a configuration as "tainted" (a taint is what signals the controller that a config is failing and a rollback should be done) it is required that all the pods trying to apply the update fail. There are rare accasions where you could have some Pods accepting the config update and some others failing, and in this case a human operator should look into the problem and decide the best way to proceed. This is why a rate of 100% failure is required for the self-healing functionality to kick in.
+- In the case you have some kind of runtime issue that somehow ends up marking a configuration as "tainted" and after solving the problem you want to reapply the same config, you need to delete the specific config revision that was marked with the taint because MARIN3R never marks a config revision back as healthy once it has been marked as tainted. To do so, list all the EnvoyConfigRevision resources related with your EnvoyConfig and delete the tainted one. You can list the EnvoyConfigRevision resources for a given node-id using `kubectl get envoyconfigrevisions -l marin3r.3scale.net/node-id=<node-id>`
```

Comment:
```suggestion
- In the case you have some kind of runtime issue that somehow ends up marking a configuration as "tainted" and after solving the problem you want to reapply the same config, you need to delete the specific config revision that was marked with the taint because MARIN3R never marks a config revision back as healthy once it has been marked as tainted. To do so, list all the EnvoyConfigRevision resources related to your EnvoyConfig and delete the tainted one. You can list the EnvoyConfigRevision resources for a given node-id using `kubectl get EnvoyConfigRevisions -l marin3r.3scale.net/node-id=<node-id>`
```